src/daft-local-execution/src/sinks/cross_join.rs

```diff
@@ -0,0 +1,119 @@
+use std::sync::Arc;
+
+use common_runtime::RuntimeRef;
+use daft_core::{join::JoinSide, prelude::SchemaRef};
+use daft_micropartition::MicroPartition;
+use tracing::instrument;
+
+use super::streaming_sink::{
+    StreamingSink, StreamingSinkExecuteResult, StreamingSinkFinalizeResult, StreamingSinkOutput,
+    StreamingSinkState,
+};
+use crate::{
+    dispatcher::{DispatchSpawner, UnorderedDispatcher},
+    ExecutionRuntimeContext,
+};
+
+#[derive(Default)]
+struct CrossJoinSinkState {
+    right_side_buffer: Vec<Arc<MicroPartition>>,
+    received_left_morsels: bool,
+    loop_index: usize,
+}
+
+impl StreamingSinkState for CrossJoinSinkState {
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+}
+
+pub struct CrossJoinSink {
+    right_schema: SchemaRef,
+}
+
+impl CrossJoinSink {
+    pub fn new(right_schema: SchemaRef) -> Self {
+        Self { right_schema }
+    }
+}
+
+impl StreamingSink for CrossJoinSink {
+    /// Cross join execute expects all right side morsels to be passed in first and then all left side morsels
+    #[instrument(skip_all, name = "CrossJoinSink::sink")]
+    fn execute(
+        &self,
+        input: Arc<MicroPartition>,
+        mut state: Box<dyn StreamingSinkState>,
+        runtime_ref: &RuntimeRef,
+    ) -> StreamingSinkExecuteResult {
+        if input.schema() == self.right_schema {
```

Comment:
this is pretty spooky. This is both expensive and error prone. What both left and right have the same schema?

Reply:
This also will be very slow since the cross join can't be parallel.

Instead I would follow what we do for the HashJoin. You can break this into 2 nodes. First do a Sink Node where you collect all the right side and then emit that to an `IntermediateNode` that can perform the cross join in parallel

Reply:
@colin-ho can help with this

Reply:
Colin actually suggested this approach. Left and right cannot have the same schema, in fact they cannot even have any common column names since we insert a project to rename the right side to remove duplicates. However I do agree that it is a little spooky since this is an assumption that we are making

@colin-ho what do you think?

Reply:
Makes sense, the `sink` -> `intermediate op` approach can enable parallel join. Check out the `ProbeStateBridge` struct, which is what is used to transfer the probe table from a build sink to the probe op. You can probably implement smth similar.

Reply:
I took a stab at implementing this, @colin-ho could you take a look at it? 

Also, would there be any value in making the intermediate op into a streaming sink that outputs the join between one table in the input and one in the collected tables at a time? That way, the join outputs can be streamed out in smaller parts at a time.

Reply:
@kevinzwang left a comment on this in my review